### PR TITLE
BUG,MAINT: Ensure masked elements can be tested against nan and inf.

### LINKED ISSUE #11121: Incorrect behavior of np.testing.assert_array_almost_equal on linux, windows
I get the following unexpected exception / test failure on linux (CentOS-6) and Windows but not on Mac.  This was driven by a CI failure that showed up for linux / windows but no on my local Mac version.
```
In [62]: c = np.ma.MaskedArray([3.4, 4.1, 6.3], mask=[False, True, False])

In [63]: np.testing.assert_array_almost_equal(c[:2], [3.4, np.nan])

In [64]: np.testing.assert_array_almost_equal(c, [3.4, np.nan, 6.3])
/export/tom/anaconda/envs/astropy/lib/python3.6/site-packages/numpy/core/arrayprint.py:686: UserWarning: Warning: converting a masked element to nan.
  s = self.format % x
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
/export/tom/anaconda/envs/astropy/lib/python3.6/site-packages/numpy/testing/utils.py in assert_array_compare(comparison, x, y, err_msg, verbose, header, precision, equal_nan, equal_inf)
    760 
--> 761         val = comparison(x, y)
    762 

/export/tom/anaconda/envs/astropy/lib/python3.6/site-packages/numpy/testing/utils.py in compare(x, y)
    952         y = array(y, dtype=dtype, copy=False, subok=True)
--> 953         z = abs(x - y)
    954 

/export/tom/anaconda/envs/astropy/lib/python3.6/site-packages/numpy/ma/core.py in __sub__(self, other)
   4078             return NotImplemented
-> 4079         return subtract(self, other)
   4080 

/export/tom/anaconda/envs/astropy/lib/python3.6/site-packages/numpy/ma/core.py in __call__(self, a, b, *args, **kwargs)
   1045             np.seterr(divide='ignore', invalid='ignore')
-> 1046             result = self.f(da, db, *args, **kwargs)
   1047         # Get the mask for the result

ValueError: operands could not be broadcast together with shapes (3,) (2,) 

During handling of the above exception, another exception occurred:

ValueError                                Traceback (most recent call last)
<ipython-input-64-2c7b42e22a4d> in <module>()
----> 1 np.testing.assert_array_almost_equal(c, [3.4, np.nan, 6.3])

/export/tom/anaconda/envs/astropy/lib/python3.6/site-packages/numpy/testing/utils.py in assert_array_almost_equal(x, y, decimal, err_msg, verbose)
    960     assert_array_compare(compare, x, y, err_msg=err_msg, verbose=verbose,
    961              header=('Arrays are not almost equal to %d decimals' % decimal),
--> 962              precision=decimal)
    963 
    964 

/export/tom/anaconda/envs/astropy/lib/python3.6/site-packages/numpy/testing/utils.py in assert_array_compare(comparison, x, y, err_msg, verbose, header, precision, equal_nan, equal_inf)
    784         msg = build_err_msg([x, y], err_msg, verbose=verbose, header=header,
    785                             names=('x', 'y'), precision=precision)
--> 786         raise ValueError(msg)
    787 
    788 

ValueError: 
error during assertion:

Traceback (most recent call last):
  File "/export/tom/anaconda/envs/astropy/lib/python3.6/site-packages/numpy/testing/utils.py", line 761, in assert_array_compare
    val = comparison(x, y)
  File "/export/tom/anaconda/envs/astropy/lib/python3.6/site-packages/numpy/testing/utils.py", line 953, in compare
    z = abs(x - y)
  File "/export/tom/anaconda/envs/astropy/lib/python3.6/site-packages/numpy/ma/core.py", line 4079, in __sub__
    return subtract(self, other)
  File "/export/tom/anaconda/envs/astropy/lib/python3.6/site-packages/numpy/ma/core.py", line 1046, in __call__
    result = self.f(da, db, *args, **kwargs)
ValueError: operands could not be broadcast together with shapes (3,) (2,) 


Arrays are not almost equal to 6 decimals
 x: MaskedArray([ 3.4,  nan,  6.3])
 y: array([ 3.4,  6.3])
```

### PR DESCRIPTION
**Pull request by mhvk**
The removal of nan and inf from arrays that are compared using test routines like assert_array_equal treated the two arrays separately, which for masked arrays meant that some elements would not be removed when they should have been. This PR corrects this.

fixes #11121

### CODE REVIEW

File: numpy/testing/_private/utils.py [hunk 1]
```diff
@@ -695,9 +695,14 @@ def isnumber(x):
     def istime(x):
         return x.dtype.char in "Mm"
 
-    def chk_same_position(x_id, y_id, hasval='nan'):
-        """Handling nan/inf: check that x and y have the nan/inf at the same
-        locations."""
+    def func_same_pos(func, x, y, hasval='nan'):
```
Comment (charris) [inline numpy/testing/_private/utils.py]: Needs a full docstring. I think the original name is a bit better. Maybe change the signature to `(x, y, func, hasval='nan')`. Do we really need both `hasval` and `func`?

File: numpy/testing/_private/utils.py [hunk 2]
```diff
@@ -706,6 +711,16 @@ def chk_same_position(x_id, y_id, hasval='nan'):
...
+        return x_id if x_id.shape else y_id
+
+    def remove_flagged(x, y, flag):
+        if flag.all():
+            return array(()), array(())
+        if x.shape:
+            x = x[~flag]
+        if y.shape:
+            y = y[~flag]
+        return x, y
```
Comment (eric-wieser) [inline numpy/testing/_private/utils.py]: I don't understand this function - can't it just be written `return x[~flag], y[~flag]`?

File: numpy/testing/_private/utils.py [hunk 3]
```diff
@@ -719,32 +734,23 @@ def chk_same_position(x_id, y_id, hasval='nan'):
...
+                xy_isinf |= func_same_pos(lambda xy: xy == -inf, x, y,
+                                          hasval='-inf')
+
+            if xy_isnan is not False:
+                if xy_isinf is not False:
+                    x, y = remove_flagged(x, y, xy_isinf | xy_isnan)
+                else:
+                    x, y = remove_flagged(x, y, xy_isnan)
+            elif xy_isinf is not False:
+                x, y = remove_flagged(x, y, xy_isinf)
```
Comment (eric-wieser) [inline numpy/testing/_private/utils.py]: Seems like micro-optimization here - can't we just write in place of these 7 lines:
```
to_remove = xy_isinf | xy_isnan
x = x[~to_remove]
y = y[~to_remove]
```

File: numpy/testing/_private/utils.py (hunk 2 above)
Comment (mhvk) [reply numpy/testing/_private/utils.py]: No, `x` or `y` can be a scalar while the other is an array. This is why the flags were treated separately before.

File: numpy/testing/_private/utils.py (hunk 3 above)
Comment (mhvk) [reply numpy/testing/_private/utils.py]: Doesn't work for the same reason.

File: numpy/testing/_private/utils.py (hunk 1 above)
Comment (mhvk) [reply numpy/testing/_private/utils.py]: `func` does not always have a name (used with a `lambda` below to test for +/- inf). I'll change to `func_chk_same_pos`.

File: numpy/testing/_private/utils.py (hunk 3 above)
Comment (mhvk) [reply numpy/testing/_private/utils.py]: The problem was that `xy_isinf` could just be `False`, in which case the flags would not broadcast correctly. But of course, this is easily checked in `remove_flagged`, so I did that now.

**Comment by mhvk** [comment]
@charris, @eric-wieser - thanks for the comments; I made some changes which I think make the use clearer.

File: numpy/testing/_private/utils.py (hunk 2 above)
Comment (eric-wieser) [reply numpy/testing/_private/utils.py]: The issue here is avoiding `arr_0d[bool_nd]`, right? Would `x, y = np.broadcast_arrays(x, y)` be off the table?

File: numpy/testing/_private/utils.py [hunk 4]
```diff
@@ -706,6 +711,19 @@ def chk_same_position(x_id, y_id, hasval='nan'):
...
                                 names=('x', 'y'), precision=precision)
             raise AssertionError(msg)
+        return x_id if x_id.shape else y_id
+
+    def remove_flagged(x, y, flag):
+        """Remove elements in x and y, noting that either may be scalar."""
+        if flag is False:
+            return x, y
+        if flag.all():
+            return array(()), array(())
```
Comment (eric-wieser) [inline numpy/testing/_private/utils.py]: nit: I'd use `[]` here for a size-0 array - I'm used to seeing `()` for a 0dim array, so this is a little jarring

File: numpy/testing/_private/utils.py [hunk 5]
```diff
@@ -706,6 +711,19 @@ def chk_same_position(x_id, y_id, hasval='nan'):
                                 % (hasval), verbose=verbose, header=header,
                                 names=('x', 'y'), precision=precision)
             raise AssertionError(msg)
+        return x_id if x_id.shape else y_id
+
+    def remove_flagged(x, y, flag):
+        """Remove elements in x and y, noting that either may be scalar."""
```
Comment (eric-wieser) [inline numpy/testing/_private/utils.py]: python or numpy scalar?

File: numpy/testing/_private/utils.py [hunk 6]
```diff
@@ -695,9 +695,14 @@ def isnumber(x):
     def istime(x):
         return x.dtype.char in "Mm"
 
-    def chk_same_position(x_id, y_id, hasval='nan'):
-        """Handling nan/inf: check that x and y have the nan/inf at the same
-        locations."""
+    def func_chk_same_pos(x, y, func=isnan, hasval='nan'):
```
Comment (eric-wieser) [inline numpy/testing/_private/utils.py]: Can we spell this `check` instead of `chk`? Or even `assert_same_predicate_pos`?

File: numpy/testing/_private/utils.py (hunk 2 above)
Comment (mhvk) [reply numpy/testing/_private/utils.py]: Not crazy, but I'd rather stick with what I have here...

File: numpy/testing/_private/utils.py (hunk 4 above)
Comment (mhvk) [reply numpy/testing/_private/utils.py]: OK

File: numpy/testing/_private/utils.py (hunk 5 above)
Comment (mhvk) [reply numpy/testing/_private/utils.py]: numpy, I'll add to the comment.

File: numpy/testing/_private/utils.py (hunk 6 above)
Comment (mhvk) [reply numpy/testing/_private/utils.py]: OK, `assert_same_pos`

File: numpy/testing/_private/utils.py [hunk 7]
```diff
@@ -706,6 +711,20 @@ def chk_same_position(x_id, y_id, hasval='nan'):
                                 % (hasval), verbose=verbose, header=header,
                                 names=('x', 'y'), precision=precision)
             raise AssertionError(msg)
+        return x_id if x_id.shape else y_id
```
Comment (eric-wieser) [inline numpy/testing/_private/utils.py]: I think this is backward - you should return the scalar boolean mask here, not the array one (which would at this point be equal to the larger array anyway). That allows you to remove all of your special cases in `remove_flagged`

File: numpy/testing/_private/utils.py [hunk 8]
```diff
@@ -719,32 +738,19 @@ def chk_same_position(x_id, y_id, hasval='nan'):
...
-                x = x[~(x_isnan | x_isinf)]
-                y = y[~(y_isnan | y_isinf)]
-            elif has_nan:
-                x = x[~x_isnan]
-                y = y[~y_isnan]
-            elif has_inf:
-                x = x[~x_isinf]
-                y = y[~y_isinf]
+                xy_isinf = func_assert_same_pos(x, y,
+                                                func=lambda xy: xy == +inf,
```
Comment (eric-wieser) [inline numpy/testing/_private/utils.py]: `np.isposinf`?

Comment (mhvk) [reply numpy/testing/_private/utils.py]: Learn something new every day...

Comment (mhvk) [reply numpy/testing/_private/utils.py]: But doesn't work for complex numbers...

File: numpy/testing/_private/utils.py (hunk 7 above)
Comment (mhvk) [reply numpy/testing/_private/utils.py]: Yes, indeed, you are completely right - that gets rid of `remove_flagged` altogether. Now implemented.

File: numpy/testing/_private/utils.py (hunk 8 above)
Comment (eric-wieser) [reply numpy/testing/_private/utils.py]: Does what you have here work for complex numbers?

File: numpy/testing/_private/utils.py [hunk 9]
```diff
@@ -706,6 +711,22 @@ def chk_same_position(x_id, y_id, hasval='nan'):
                                 % (hasval), verbose=verbose, header=header,
                                 names=('x', 'y'), precision=precision)
             raise AssertionError(msg)
+        # If there is a scalar, then here we know the array has the same
+        # flag as it everywhere, so we should return the scalar flag.
+        return x_id if not x_id.shape else y_id
+
+    def remove_flagged(x, y, flag):
```
Comment (eric-wieser) [inline numpy/testing/_private/utils.py]: Don't forget to remove this function now

File: numpy/testing/_private/utils.py [hunk 10]
```diff
@@ -718,49 +739,32 @@ def chk_same_position(x_id, y_id, hasval='nan'):
...
-                x_isnat, y_isnat = isnat(x), isnat(y)
+                flagged = func_assert_same_pos(x, y, func=isnat, hasval="NaT")
 
-                if any(x_isnat) or any(y_isnat):
-                    chk_same_position(x_isnat, y_isnat, hasval="NaT")
-
-                if any(x_isnat) or any(y_isnat):
-                    x = x[~x_isnat]
-                    y = y[~y_isnat]
+        if getattr(flagged, 'shape', ()):
```
Comment (eric-wieser) [inline numpy/testing/_private/utils.py]: Isn't this attribute guaranteed to exist?

**Comment by eric-wieser** [comment]
Does the inf-checking behave on complex numbers?

File: numpy/testing/_private/utils.py (hunk 10 above)
Comment (mhvk) [reply numpy/testing/_private/utils.py]: No, it can be `False` -- initialized as such. I tried changing, but cannot get rid of the `if flagged` stanza, nor of the `isinstance` - I suggest we leave this to another PR...

**Comment by mhvk** [comment]
On the `inf` checking on complex numbers: I really don't want to go there right now...  Am having too much fun with the `gufunc` signature stuff..

File: numpy/testing/_private/utils.py (hunk 10 above)
Comment (eric-wieser) [reply numpy/testing/_private/utils.py]: Just initialize it to `flagged = np.False_`?

**Comment by eric-wieser** [comment]
Complex number checking is out of scope here - that's fine. Maybe open an issue about it if you can produce a test-case

File: numpy/testing/_private/utils.py (hunk 10 above)
Comment (mhvk) [reply numpy/testing/_private/utils.py]: I thought about that, but I do feel that I'm then decreasing the clarity of the rest of the code - at least in my emacs, `False` is coloured very nicely, and `false_` is not.

**Comment by mhvk** [comment]
Cannot immediately produce one. Partially because of an oddity:
```
1.+1j*np.inf
# (nan+infj)
```
But even accounting for that, no, I cannot seem to break it!

File: numpy/testing/_private/utils.py (hunk 10 above)
Comment (eric-wieser) [reply numpy/testing/_private/utils.py]: I'd argue that keeping a variable type consistent is more important than having the color look like in your editor.

> I tried changing, but cannot get rid of the if flagged stanza

I'm gonna do a checkout to see what I'm missing here

**Comment by eric-wieser** [comment]
Interesting oddity there. I suppose you have to use `complex(1, np.inf)`

File: numpy/testing/_private/utils.py (hunk 10 above)
Comment (eric-wieser) [reply numpy/testing/_private/utils.py]: How about `bool_(False)`, which meets both requirements?

Comment (eric-wieser) [reply numpy/testing/_private/utils.py]: Ok, I see why the 0d case still has special handling - I'd forgotten that `arr[bool_0d].shape == (?,) + arr.shape`, and is not 1D.

File: numpy/testing/_private/utils.py [hunk 11]
```diff
@@ -706,6 +711,9 @@ def chk_same_position(x_id, y_id, hasval='nan'):
                                 % (hasval), verbose=verbose, header=header,
                                 names=('x', 'y'), precision=precision)
             raise AssertionError(msg)
+        # If there is a scalar, then here we know the array has the same
+        # flag as it everywhere, so we should return the scalar flag.
+        return x_id if not x_id.shape else y_id
```
Comment (eric-wieser) [inline numpy/testing/_private/utils.py]: nit: I think this is clearer as `x_id.ndim == 0` rather than `not x_id.shape`

File: numpy/testing/_private/utils.py (hunk 10 above)
Comment (eric-wieser) [reply numpy/testing/_private/utils.py]: Or an alternate spelling - `np.ndim(flagged) == 0`

**Comment by eric-wieser** [comment]
Just the nits about `ndim` and `shape` remain - you're right that it's harder to remove the special cases than I'd though

File: numpy/testing/_private/utils.py (hunk 10 above)
Comment (mhvk) [reply numpy/testing/_private/utils.py]: Yes, I like the `ndim` - that is much clearer why we do it. Pushed up.

File: numpy/testing/_private/utils.py (hunk 11 above)
Comment (mhvk) [reply numpy/testing/_private/utils.py]: Yes, done.

**Comment by eric-wieser** [comment]
Took the liberty of adding and explicit `!= 0` comparison, since that's normally how we spell ndim comparisons

**Review by eric-wieser — APPROVED** [review]

File: numpy/testing/_private/utils.py [hunk 12]
```diff
@@ -695,9 +695,14 @@ def isnumber(x):
...
-    def chk_same_position(x_id, y_id, hasval='nan'):
-        """Handling nan/inf: check that x and y have the nan/inf at the same
-        locations."""
+    def func_assert_same_pos(x, y, func=isnan, hasval='nan'):
+        """Handling nan/inf: combine results of running func on x and y,
+        checking that they are True at the same locations."""
+        x_id = func(x)
+        y_id = func(y)
+        if not any(x_id) and not any(y_id):
+            return False
```
Comment (eric-wieser) [inline numpy/testing/_private/utils.py]: Are these two lines needed? Aren't they covered by the below?

Comment (mhvk) [reply numpy/testing/_private/utils.py]: You'd think not, but numerous recursion failures if I leave them out...

Comment (mhvk) [reply numpy/testing/_private/utils.py]: So, what happens is that it does recurse, but the `if` statement prevents it from doing it more than once.

Comment (mhvk) [reply numpy/testing/_private/utils.py]: which is simple to solve - just use `any` - will push fix up shortly.

Comment (mhvk) [reply numpy/testing/_private/utils.py]: Doesn't work either, but passing in `equal_[nan|inf]=False` does.

Comment (mhvk) [reply numpy/testing/_private/utils.py]: No, it doesn't...

Comment (eric-wieser) [reply numpy/testing/_private/utils.py]: Maybe add a comment saying that this is needed to prevent recursion or something?

Good to go either way

**Comment by mhvk** [comment]
What a pain! But finally got it working with a more sensible implementation. Turns out I was partially misled by test that failed because they used subclasses that were not written correctly (one by me...).

File: numpy/testing/_private/utils.py [hunk 13]
```diff
@@ -695,17 +695,23 @@ def isnumber(x):
...
-    def chk_same_position(x_id, y_id, hasval='nan'):
-        """Handling nan/inf: check that x and y have the nan/inf at the same
-        locations."""
-        try:
-            assert_array_equal(x_id, y_id)
-        except AssertionError:
+    def func_assert_same_pos(x, y, func=isnan, hasval='nan'):
+        """Handling nan/inf: combine results of running func on x and y,
+        checking that they are True at the same locations."""
+        # Both the cast to bool_ and the != True comparison are done
```
Comment (eric-wieser) [inline numpy/testing/_private/utils.py]: Can you move the cast to bool to the scalar case below?

File: numpy/lib/tests/test_ufunclike.py [hunk 14]
```diff
@@ -55,6 +55,10 @@ def __array_wrap__(self, obj, context=None):
                 obj.metadata = self.metadata
                 return obj
 
+            def __array_finalize__(self, obj):
+                self.metadata = getattr(obj, 'metadata', None)
```
Comment (eric-wieser) [inline numpy/lib/tests/test_ufunclike.py]: Can you describe in the commit message why these changes were needed?

File: numpy/testing/_private/utils.py (hunk 13 above)
Comment (mhvk) [reply numpy/testing/_private/utils.py]: I needed to cast both `x_id` and `y_id`, but that can indeed be done at the end. That does look a bit nicer, so done.

File: numpy/lib/tests/test_ufunclike.py (hunk 14 above)
Comment (mhvk) [reply numpy/lib/tests/test_ufunclike.py]: Done. This was indeed a non-trivial one - I had to look again to remind me why it was needed.  I wondered about the use of stating it in the commit message until I recalled that with `git blame` one can find why this was added.

File: numpy/testing/_private/utils.py [hunk 15]
```diff
@@ -695,17 +695,24 @@ def isnumber(x):
...
+        if (x_id == y_id).all() != True:
             msg = build_err_msg([x, y],
                                 err_msg + '\nx and y %s location mismatch:'
                                 % (hasval), verbose=verbose, header=header,
                                 names=('x', 'y'), precision=precision)
             raise AssertionError(msg)
+        # If there is a scalar, then here we know the array has the same
+        # flag as it everywhere, so we should return the scalar flag.
+        return (bool_(x_id) if x_id.ndim == 0 else
+                (bool_(y_id) if y_id.ndim == 0 else y_id))
```
Comment (eric-wieser) [inline numpy/testing/_private/utils.py]: Probably clearer as just a normal if / else chain at this point.

File: numpy/testing/_private/utils.py [hunk 16]
```diff
@@ -718,49 +725,32 @@ def chk_same_position(x_id, y_id, hasval='nan'):
...
-                if any(x_isnat) or any(y_isnat):
-                    x = x[~x_isnat]
-                    y = y[~y_isnat]
+        if flagged.ndim > 0:
+            x, y = x[~flagged], y[~flagged]
+            # Only do the comparison if actual values are left
+            if x.size == 0:
+                return
+        elif flagged:
+            # no sense doing comparison if everything is flagged.
```
Comment (eric-wieser) [inline numpy/testing/_private/utils.py]: Not sure I agree with this - what if `x` and `y` are quantities of different dimensions? The test should still compare the empty arrays so that the correct error is raised, right?

**Comment by eric-wieser** [comment]
Sorry for dropping the ball on this one. Would be great to get this into 1.15.

File: numpy/testing/_private/utils.py (hunk 16 above)
Comment (mhvk) [reply numpy/testing/_private/utils.py]: It is the same case as the `if x.size == 0` stanza above, but for the case where `x` and `y` are both scalar. In principle, I agree the comparison should work for empty elements, but in practice, this is not the case (e.g., `matrix` changing its shape being one problem...)

Since this was here before, I propose to leave it as is.

File: numpy/testing/_private/utils.py (hunk 15 above)
Comment (mhvk) [reply numpy/testing/_private/utils.py]: Agreed; done.

File: numpy/testing/_private/utils.py (hunk 16 above)
Comment (eric-wieser) [reply numpy/testing/_private/utils.py]: Sure, i agree this is out of scope for the patch

**Review by eric-wieser — APPROVED** [review]

File: numpy/testing/_private/utils.py (hunk 16 above)
Comment (mhvk) [reply numpy/testing/_private/utils.py]: Overall, it sadly remains a chunk of rather fragile code...

**Comment by charris** [comment]
Thanks Marten.

File: numpy/testing/_private/utils.py [hunk 17]
```diff
@@ -695,17 +695,28 @@ def isnumber(x):
...
-        except AssertionError:
+    def func_assert_same_pos(x, y, func=isnan, hasval='nan'):
+        """Handling nan/inf: combine results of running func on x and y,
+        checking that they are True at the same locations."""
+        # Both the != True comparison here and the cast to bool_ at
+        # the end are done to deal with `masked`, which cannot be
+        # compared usefully, and for which .all() yields masked.
+        x_id = func(x)
+        y_id = func(y)
+        if (x_id == y_id).all() != True:
```
Comment (charris) [inline numpy/testing/_private/utils.py]: Hmm,
```
In [10]: ma.masked != True
Out[10]: masked

In [11]: ma.masked == True
Out[11]: masked

In [12]: bool(ma.masked)
Out[12]: False
```

Comment (mhvk) [reply numpy/testing/_private/utils.py]: The purpose is to ensure that the stanza is *not* executed if `x_id` or `y_id` is `masked`, yet is executed if they are different.
